### DISCUSSION TITLE
Time-saving file templates of Illuminated Cloud 2 to create some sets of Apex Classes of different Layers

### DISCUSSION THREAD

**Discussion by PeterLin888** [apex-enterprise-patterns/fflib-apex-common#343] [Q&A] [answered]
Hi AEP Community, especially those who are using JetBrains IDEs,

CC: @wimvelzeboer @SCWells72

I was looking for some time-saving file templates of Illuminated Cloud 2 (IC2) to create some sets of Apex Classes of different Layers and deploy them together.

Because I was not able to find anything about these, I created one set at [aep-ic2-templates-jetbrains-webstorm-intellij-idea](https://github.com/aepsf/aep-ic2-templates-jetbrains-webstorm-intellij-idea). This set of templates require manually naming of all the class names. I hope I can find some ways to improve the UI, probably something like IC2's "Create New Apex Trigger" window to automatically populate all class names and configuration based on the SObject type and allow us to change the auto populated class names like the IC2's "Create New Lightning Web Component" window.

Could you please share with me if you happen to know any similar time-saving file templates (Illuminated Cloud 2)?

Thanks,
Peter

**Comment by wimvelzeboer** [comment]
@PeterLinBiz I created something similar a while ago for other developers in my company, but not that advanced. 
It had a separate template for each file, and could not do interfaces.

It would be great if @SCWells72 can add something to Illuminated Cloud that would make it more smooth. 
The naming convention is very clear so you should be able to automatically create the interface and implementation structure for Domain, Selectors, Services and TriggerHandlers.

Another thing I would like to add (if possible) is automatically use the Falcon-template folders when creating templates; 
like here: https://github.com/apex-enterprise-patterns/fflib-apex-common-samplecode/tree/master/sfdx-source/apex-common-samplecode/main/classes 

And finally on my wish-list for @SCWells72 is the ability to add method signatures on interfaces for methods created on implementation classes. Or the other way around for Service methods signatures to create methods in the implementation and their static class.

**Comment by SCWells72** [comment, marked as answer]
This is great stuff, @PeterLinBiz. Regarding how/whether it might be possible to have something like a file template work for a collection of related files vs. a single file, I'll have to investigate. I'll use the linked IC2 BitBucket issue for that work. Ideally it'd be something offered directly in the JetBrains plugin SDK vs. something that would have to be home-grown, but we'll see.

@wimvelzeboer regarding your wish list item, part of it sounds like an Extract Interface refactoring which is on my relatively short-term TODO list; part of it sounds like the existing functionality to implement/override methods from base classes/interfaces; finally a third part sounds kind of like the latter but for stubbing class methods according to a pattern? Is that right? If that's a correct assessment, the first is already on my radar; the second should already be done; feel free to log an enhancement for the third explaining what you're looking to accomplish. Thanks!
